bootstrap/env.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"hash/fnv"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+// SetDefaultEnv does os.Setenv(key, value) if key does not exist (os.LookupEnv)
+// It returns true if the key was set
+func SetDefaultEnv(key, value string) (bool, error) {
+	_, exists := os.LookupEnv(key)
+	if !exists {
+		return true, os.Setenv(key, value)
+	}
+	return false, nil
+}
+
+// EnvEqual returns true if both keys have the same value or both keys do not
+// exist. If the values are different or if one key is "" and the other
+// is not set it will return false.
+func EnvEqual(key1, key2 string) bool {
+	val1, exists1 := os.LookupEnv(key1)
+	val2, exists2 := os.LookupEnv(key2)
+	return val1 == val2 && exists1 == exists2
+}
+
+// SetupMagicEnviroment sets magic environment variables scripts currently expect.
+func SetupMagicEnviroment(job string) (err error) {
+	home := os.Getenv(HomeEnv)
+	/*
+		TODO(fejta): jenkins sets these values. Consider migrating to using
+					 a secret volume instead and passing the path to this volume
+					into bootstrap.py as a flag.
+	*/
+	_, err = SetDefaultEnv(
+		GCEPrivKeyEnv,
+		filepath.Join(home, ".ssh/google_compute_engine"),
+	)
+	if err != nil {
+		return err
+	}
+	_, err = SetDefaultEnv(
+		GCEPubKeyEnv,
+		filepath.Join(home, ".ssh/google_compute_engine.pub"),
+	)
+	if err != nil {
+		return err
+	}
+	_, err = SetDefaultEnv(
+		AWSPrivKeyEnv,
+		filepath.Join(home, ".ssh/kube_aws_rsa"),
+	)
+	if err != nil {
+		return err
+	}
+	_, err = SetDefaultEnv(
+		AWSPubKeyEnv,
+		filepath.Join(home, ".ssh/kube_aws_rsa.pub"),
+	)
+	if err != nil {
+		return err
+	}
+
+	// TODO(bentheelder): determine if we can avoid getcwd here :-/
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	/*
+		 TODO(fejta): jenkins sets WORKSPACE and pieces of our infra expect this
+					  value. Consider doing something else in the future.
+		 Furthermore, in the Jenkins and Prow environments, this is already set
+		 to something reasonable, but using cwd will likely cause all sorts of
+		 problems. Thus, only set this if we really need to.
+	*/
+	_, err = SetDefaultEnv(WorkspaceEnv, cwd)
+	if err != nil {
+		return err
+	}
+	/*
+	 By default, Jenkins sets HOME to JENKINS_HOME, which is shared by all
+	 jobs. To avoid collisions, set it to the cwd instead, but only when
```

Comment:
How does this avoid collisions? Does each job start in a different directory?

Reply:
I think so, this is the exact behavior from jenkins/bootstrap.py:
https://github.com/kubernetes/test-infra/blob/604f6714528c6ddad49b248230f8d7b03a7e641e/jenkins/bootstrap.py#L746-L750

Reply:
confirmed with Jeff (who wrote this in bootstrap.py), each job starts in a different directory.